jekyll.bzl

```diff
@@ -0,0 +1,110 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def _bucket_from_worspace_name(wname):
+  revlist = []
+  for part in wname.split("_"):
+    revlist.insert(0, part)
+  return ".".join(revlist)
+
+def _impl(ctx):
+  """Quick and non-hermetic rule to build a Jekyll site."""
+  source = ctx.actions.declare_directory(ctx.attr.name + "-srcs")
+  output = ctx.actions.declare_directory(ctx.attr.name + "-build")
+
+  ctx.actions.run_shell(inputs = ctx.files.srcs,
+    outputs = [source],
+    command = ("mkdir -p %s\n" % (source.path)) +
+    "\n".join([
+      "tar xf %s -C %s" % (src.path, source.path) for src in ctx.files.srcs])
+  )
+  ctx.actions.run(
+    inputs = [source],
+    outputs = [output],
+    executable = "jekyll",
+    use_default_shell_env = True,
+    arguments = ["build", "-q", "-s", source.path, "-d", output.path]
+  )
+  ctx.actions.run(
+    inputs = [output],
+    outputs = [ctx.outputs.out],
+    executable = "tar",
+    arguments = ["cf", ctx.outputs.out.path, "-C", output.path, "."]
+  )
+
+  # Create a shell script to serve the site locally or push with the --push
+  # flag.
+  bucket = ctx.attr.bucket if ctx.attr.bucket else _bucket_from_worspace_name(ctx.workspace_name)
+
+  ctx.actions.write(
+    ctx.outputs.executable,
+    """#!/bin/sh
+HOST="${HOST-localhost}"
+PORT="${PORT-12345}"
+RUNFILES=$(cd ${JAVA_RUNFILES-$0.runfiles}/%s && pwd -P)
+SOURCE_DIR="$RUNFILES/%s"
+prod_dir="$RUNFILES/%s"
+bucket="%s"
+
+serve() {
+  TDIR=$(mktemp -d)
+  RDIR=$(mktemp -d)
+  trap "rm -fr $RDIR $TDIR" EXIT
+  (cd $RDIR && \
+    jekyll serve --host "$HOST" --port "$PORT" -s "$SOURCE_DIR" -d "$TDIR")
+}
+
+push() {
+  # Get gsutil
+  gs="${GSUTIL:-$(which gsutil 2>/dev/null || true) -m}"
```

Comment:
What is `-m` in the end?
Also, `true` *is* an executable (`/usr/bin/true`). I suggest rewriting this check like so:

```
if ! gs="$(which gsutil 2>/dev/null)"; then
  echo "..." >&2
  exit 1
fi
```

See:
```
$ if ! x=$(which bash 2>/dev/null ); then echo error; fi ; echo $x
/usr/bin/bash

$ if ! x=$(which foo 2>/dev/null ); then echo error; fi ; echo $x
error

```

Reply:
The `-m` is a mistake, will remove (it must have fallen through the cracks when I cleaned the PR).

What is the benefit of the suggestion?

Reply:
> What is the benefit of the suggestion?

It's cleaner. `true` is a binary, and although `[ -x true ]` is false, I had to actually try it to verify that it is indeed false.

Reply:
Ah, I missed using `$GSUTIL`'s value if it's checked. Then how about:

```
gs="${GSUTIL:-}"
if [[ -z "$gs" -a ! gs="$(which gsutil 2>/dev/null)" ]]; then
  ...
```

Reply:
Or something similar. I just want to get rid of setting `true` to `gs` and relying on `[ -x true ]` being false.

Reply:
`|| true` was not returning the string `true`, it was returning the empty string (but make sure the command does not fails.

I replaced it with `|| :` to be sure we use the built in command (most shell consider `true` as built-ins but according to @aehlig, `:` is the POSIX standard).